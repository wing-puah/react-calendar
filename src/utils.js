function timeFns(timeStamp) {
  const month = timeStamp.getUTCMonth();
  const year = timeStamp.getUTCFullYear();
  // const date = timeStamp.getUTCDate();
  const setFirstDateOfMth = new Date(timeStamp.setUTCDate(1));
  const getFirstDayOfMth = setFirstDateOfMth.getUTCDay();
  const getLastDateOfMth = new Date(Date.UTC(year, month+1, 0));
  const getLastDayOfMth = getLastDateOfMth.getUTCDay();
  const toLocalStorageString = year.toString().concat(
    month > 9
      ? month
      : '0' + month.toString()
  );
  // console.log('running');
  return {
    setFirstDateOfMth,
    getFirstDayOfMth,
    selectedDay: timeStamp.getUTCDay(),
    getLastDateOfMth,
    getLastDayOfMth,
    toLocalStorageString,
  }
}

function getStartOfWeek(timeStamp) {
  const firstDay = timeFns(timeStamp).getFirstDayOfMth;
  const firstDateTimeStamp = timeFns(timeStamp).setFirstDateOfMth;

  const firstDateOfWeek = firstDay === 0 ? firstDateTimeStamp
    : new Date(firstDateTimeStamp.setDate(firstDateTimeStamp.getUTCDate() - firstDay))

  return firstDateOfWeek;
}

function getLastOfWeek(timeStamp) {
  const lastDay = timeFns(timeStamp).getLastDayOfMth;
  const lastDateTimeStamp = timeFns(timeStamp).getLastDateOfMth;

  const lastDateOfWeek = lastDay === 6 ? lastDateTimeStamp
    : new Date(lastDateTimeStamp.setDate(lastDateTimeStamp.getDate() + (7-lastDay)))

  return lastDateOfWeek
}


function sendToLocalStorage(form, func, edit) {
  let submission = {};
  const formData = new FormData(form);
  const LSlength = localStorage.length;
  let idx = LSlength > 0 ? Number(localStorage.key(LSlength-1)) +1 : 0;

  submission['index'] = idx;
  for(let entry of formData.entries()) {
    submission[entry[0]] = entry[1];
  }

  if(edit===true){
    idx = form.getAttribute('idx')
  }

  localStorage.setItem(idx, JSON.stringify(submission));

  if(func) {
    func();
  }
}


export { timeFns, getStartOfWeek, getLastOfWeek, sendToLocalStorage };
