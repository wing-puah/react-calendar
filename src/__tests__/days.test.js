import React from 'react';
import renderer from 'react-test-renderer';
import Days, { renderDays } from '../days';
import { timeFns, getStartOfWeek, getLastOfWeek } from '../utils';

describe('Days', () => {
  const timeStamp = new Date(2019, 3, 13, 14, 4, 15);
  const DaysComponent = renderer
    .create(<Days timeStamp={timeStamp}/>)
    .toJSON();

  const DaysRenderFn = renderDays(timeStamp);
  const { getLastDateOfMth, getFirstDayOfMth, getLastDayOfMth } = timeFns(timeStamp);
  const lastDateInt = getLastDateOfMth.getDate();
  const noOfDaysOnCalendar = lastDateInt + getFirstDayOfMth + (6-getLastDayOfMth);

  it('renders correctly', () => {
    expect(DaysComponent).toMatchSnapshot();
  });

  it('renders all the dates in the month', () => {
    expect(DaysRenderFn.length).toBe(Math.ceil(lastDateInt/7))
  });
})
