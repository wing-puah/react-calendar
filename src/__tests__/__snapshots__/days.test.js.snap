// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Days renders correctly 1`] = `
Array [
  <div
    className="calendar__col"
  >
    month: 
    Apr
     year: 
    2019
     last date: 
    30
     first day: 
    Mon
  </div>,
  <div
    className="calendar__header"
  >
    <div
      className="calendar__col"
    >
      Sunday
    </div>
    <div
      className="calendar__col"
    >
      Monday
    </div>
    <div
      className="calendar__col"
    >
      Tuesday
    </div>
    <div
      className="calendar__col"
    >
      Wednesday
    </div>
    <div
      className="calendar__col"
    >
      Thursday
    </div>
    <div
      className="calendar__col"
    >
      Friday
    </div>
    <div
      className="calendar__col"
    >
      Saturday
    </div>
  </div>,
  <div
    className="calendar__body"
  >
    <div
      className="calendar__row"
    >
      <div
        className="calendar__col calendar__past calendar--disabled"
      >
        31
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        1
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        2
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        3
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        4
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        5
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        6
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
    </div>
    <div
      className="calendar__row"
    >
      <div
        className="calendar__col"
      >
        7
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        8
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        9
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        10
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        11
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        12
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        13
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
    </div>
    <div
      className="calendar__row"
    >
      <div
        className="calendar__col"
      >
        14
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        15
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        16
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        17
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        18
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        19
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        20
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
    </div>
    <div
      className="calendar__row"
    >
      <div
        className="calendar__col"
      >
        21
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        22
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        23
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        24
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        25
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        26
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        27
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
    </div>
    <div
      className="calendar__row"
    >
      <div
        className="calendar__col"
      >
        28
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        29
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col"
      >
        30
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col calendar__next calendar--disabled"
      >
        1
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col calendar__next calendar--disabled"
      >
        2
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col calendar__next calendar--disabled"
      >
        3
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
      <div
        className="calendar__col calendar__next calendar--disabled"
      >
        4
        <div
          className="event"
        >
          Return this event lor.
        </div>
      </div>
    </div>
  </div>,
]
`;
